walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> options = new ArrayList();
+    ArrayList<TimeRange> unavailableBlocks = new ArrayList();
+    long duration = request.getDuration();
+
+    // Meetings can't be longer than a day
+    if (duration > 24 * 60) {
+      return options;
+    } else {
+      options.add(TimeRange.WHOLE_DAY);
+    }
+    
+    unavailableBlocks = unavailableTime(events, request);
+
+    for (TimeRange block : unavailableBlocks) {
+      int busyStart = block.start();
+      int busyEnd = block.end();
+
+      for (TimeRange freeTime : new ArrayList<>(options)) {
+
+        // Conflict in ranges
+        if (block.overlaps(freeTime)) {
+          int freeStart = freeTime.start();
+          int freeEnd = freeTime.end();
+
+          // Free time extends beyond busy time
+          if (!block.contains(freeTime)) {
+            // Check for free time before event starts
+            if ((freeStart + duration) <= busyStart) {
+              options.add(TimeRange.fromStartEnd(freeStart, busyStart, false));
+            }
+            // Check for free time after event ends
+            if ((busyEnd + duration) <= freeEnd) {
+              options.add(TimeRange.fromStartEnd(busyEnd, freeEnd, false));
+            }
+          }
+          options.remove(freeTime);
+        }
+      } 
+    }
+    return options;
+  }
+
+  /**
+   * Block off all unavailable time
+   */
+  private ArrayList<TimeRange> unavailableTime(Collection<Event> allEvents, MeetingRequest request) {
+    ArrayList<TimeRange> attendeeEvents = attendeeEvents(allEvents, request);
+    return merge(attendeeEvents, request);
+  }
+
+  /**
+   * Merge all attendees' events
+   */
+  private ArrayList<TimeRange> merge(ArrayList<TimeRange> events, MeetingRequest request) {
+    long duration = request.getDuration();
+    Collections.sort(events, TimeRange.ORDER_BY_START);
+
+    for (int i = 0; i < events.size() - 1; i++) {
+      TimeRange current = events.get(i);
+      TimeRange next = events.get(i+1);
+      boolean nextEnd = (next.end() == TimeRange.END_OF_DAY);
```

Comment:
The name `nextEnd` seems a bit ambiguous.
It seems to handle the end of day edge case, but maybe you can explain what it is in the comments?